p2p: attempt to fill full outbound connection slots with peers that support tx relay

mzumsande opened the pull request:
As described in the issues #16418 and #28371, it's a possibility that we could end up having an insufficient number of outbound peers relaying our transactions. Having fewer outbound peers support tx-relay could also have implications on [privacy](https://github.com/bitcoin/bitcoin/issues/16418#issuecomment-514328042) and fee estimation.

While #28488 is suggesting meaures based on comparing fee filters / mempool sizes, there is also the simpler issue of peers that tell us they don't want transactions, which is determined by the `fRelay` field in the `version` msg. If such a peer runs bitcoin core, that usually means that it's running in `-blocksonly` mode and accepting inbound connections. The status quo is that we will not avoid these peers and just not send them any transactions.

This PR proposes not to waste any of our 8 full-relay outbound slots on these peers (unless we are in `-blocksonly` mode ourselves). Using them as outbound peers for block-relay-only connections is fine though and not impacted.

As was suggested by ajtowns below, we don't disconnect during version processing, but later during the regularly scheduled `EvictExtraOutboundPeers()` task, and only if we are at the maximum of full-outbound peers.

If reviewers think that this proposal is too aggressive, an alternative solution (with a little more complexity) would be to restrict the maximum number of outbound connections to `-blocksonly` peers to 1 or 2. Although I currently think that it's ok to be selective with outbound connections, so I didn't implement that right away.

mzumsande committed:
p2p: Evict outbound connections to -blocksonly peers when at full outbound maximum

...unless we are in -blocksonly mode ourselve or in ibd.
If we had too many outbound peers not participating in tx relay, we would run
into problems with getting own transaction relayed. This would also be
bad for privacy and fee estimation.

We evict non-txrelay peers only when all full outbound slots
are filled to have peers in hypothetical situations where
-blocksonly peers are the only ones available.

Idea for this approach by ajtowns.

mzumsande committed:
test: add test for interaction with -blocksonly peers

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -3330,6 +3330,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         if (!vRecv.empty())
             vRecv >> fRelay;
+
+        // Don't waste any of our full inbound slots on peers that don't want tx relay (because they might be in -blocksonly mode)
```
`inbound` -> `outbound`

mzumsande replied:
oops - fixed!

naumenkogs reviewed (approved):
Concept ACK.

I think this measure makes perfect sense, it's easy and immediately approaches the mentioned issues.

I would be fine with restricting this (as suggested by the OP message) too.

brunoerg commented on src/net_processing.cpp:
```diff
@@ -3330,6 +3330,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         if (!vRecv.empty())
             vRecv >> fRelay;
+
+        // Don't waste any of our full outbound slots on peers that don't want tx relay (because they might be in -blocksonly mode)
```
In 82e7a7dc3b4447e4affe3a98e1af8b478aa4dc04: Perhaps we could mention on this comment that it doesn't apply in case we're running in `-blocksonly` mode.
```suggestion
        // Don't waste any of our full outbound slots on peers that don't want tx relay (because they might be in -blocksonly mode) unless we are in -blocksonly mode.
```

mzumsande replied:
I have included this info in the doc for the new spot!

ajtowns commented on src/net_processing.cpp:
```diff
@@ -3330,6 +3330,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        // Don't waste any of our full outbound slots on peers that don't want tx relay (because they might be in -blocksonly mode)
+        if (!fRelay && !m_opts.ignore_incoming_txs && pfrom.IsFullOutboundConn())
+        {
+            LogPrint(BCLog::NET, "peer=%d (full-relay outbound) does not participate in transaction relay; disconnecting\n", pfrom.GetId());
```
Rather than doing this immediaely upon connection (ie, after receiving their VERSION message), would it be better to do it only when we have all our outbound slots filled up (ie, as part of `EvictExtraOutboundPeers`) ?

Also, doing the check repeatedy would mean that we could keep blocks-only peers while we're doing IBD, only dropping them when we exit IBD and start caring about tx relay.

mzumsande replied:
> Rather than doing this immediaely upon connection (ie, after receiving their VERSION message), would it be better to do it only when we have all our outbound slots filled up (ie, as part of EvictExtraOutboundPeers) ?

I like that idea. If for some strange reason we wouldn't be able to find any other peers we wouldn't end up connecting the only ones we have. But I think we would want to start evicting if we are at the limit (so in `CheckForStaleTipAndEvictPeers`), not just if we have exceeded it (`EvictExtraOutboundPeers`).

> Also, doing the check repeatedy would mean that we could keep blocks-only peers while we're doing IBD, only dropping them when we exit IBD and start caring about tx relay.

Do we really want that? The main reason to run in `-blocksonly` mode is to save traffic. I would assume that most nodes wouldn't enable listening anyway, but those that do (maybe not even on purpose), maybe it would be nicer not to use them for IBD?

naumenkogs replied:
>start evicting if we are at the limit (so in CheckForStaleTipAndEvictPeers), not just if we have exceeded it (EvictExtraOutboundPeers).

I agree.

>Do we really want that? The main reason to run in -blocksonly mode is to save traffic. I would assume that most nodes wouldn't enable listening anyway, but those that do (maybe not even on purpose), maybe it would be nicer not to use them for IBD?

Not sure about this. You seem to care specifically about those users which do `-blocksonly` and forget to disable inbounds. Perhaps this is better communicated through documentation and logging. I think we should leave them a chance to strengthen the block relaying network.

mzumsande replied:
> it seems odd / logically inconsistent to run a node in -blocksonly and enable inbounds, so I don't think it's worth overly optimizing for the use case

Maybe there is a use case for wanting to help the network by providing inbound capacity, but still minimize traffic at the same time? Although then it often doesn't make a lot of sense to run a non-pruned node - unless you also need the entire chain for selfish purposes (e.g. txindex).
In any case, I have included the suggestion to not evict while in ibd in the latest push!

ajtowns replied:
> Do we really want that? The main reason to run in `-blocksonly` mode is to save traffic.

In that case you'd be better off doing `-maxuploadtarget` either instead of or in addition to `-blocksonly`?

> it seems odd / logically inconsistent to run a node in `-blocksonly` and enable inbounds,

I think it would make sense to expect that if 100% (or 99.9%) of nodes were `-blocksonly` that they would still manage to propagate blocks sensibly, bring up new nodes efficiently, etc. In theory with this PR, we would also be able to propagate txs successfully in the 99.9% `-blocksonly` case.

amitiuttarwar reviewed (commented):
concept ACK

mzumsande commented:
I've now changed the approach to @ajtowns suggestion to disconnecting during eviction.

glozow commented on src/net_processing.cpp:
```diff
@@ -5246,6 +5247,28 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
         }
     }
+
+    // If we are at the max full outbound limit but not above, check if any peers
```
nit f3b74d43823a038fb0aa4e0bcb7d4f99a7f90688:

The 2 logical paths depending on the value of `extra_outbound_count` could be more clear if this were an `if/else if` block.

mzumsande replied:
changed to `else if`

glozow commented on test/functional/test_framework/test_node.py:
```diff
@@ -673,6 +673,7 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, **kwargs):
     def add_outbound_p2p_connection(self, p2p_conn, *, wait_for_verack=True, p2p_idx, connection_type="outbound-full-relay", **kwargs):
         """Add an outbound p2p connection from node. Must be an
         "outbound-full-relay", "block-relay-only", "addr-fetch" or "feeler" connection.
+        The txrelay arg determines whether our peer will signal support for txrelay in their version message.
```
831ed4960bfb1eafb763cb2a52201a97a17f8168: seems like this should be in the `peer_connect_send_version` function docstring?

mzumsande replied:
Removed the doc, since it's hidden in the `kwargs` anyway (It wasn't when I added it in an earlier revision). Within `peer_connect_send_version` it's kind of self-explanatory, I had added it here to avoid confusion to which side of the connection `txrelay` refers to.

glozow reviewed (commented):
concept/approach ACK

sr-gi reviewed (commented):
Concept ACK, just some nits here and there

sr-gi commented on src/net_processing.cpp:
```diff
@@ -5246,6 +5247,28 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
+    // because then we don't care if our peer supports tx relay.
+    if (extra_outbound_count == 0 && !m_opts.ignore_incoming_txs && !m_chainman.IsInitialBlockDownload()) {
+        std::optional<NodeId> node_to_evict;
+        m_connman.ForEachNode([&node_to_evict](CNode* node) {
+            if (!node->IsFullOutboundConn() || node->m_relays_txs) return;
+            if (!node_to_evict.has_value() || *node_to_evict < node->GetId()) {
+                node_to_evict = node->GetId();
+            }
+        });
```
What if there is, at least, a node already flagged to be disconnected? Shouldn't we skip this altogether? Otherwise, it could be the case that we end up evicting more nodes that we really need

sr-gi replied:
Nvm, `ForEachNode` already checks `NodeFullyConnected`

mzumsande replied:
Yes, that's why I didn't include the check here (the similar checks a few lines above aren't necessary either)

naumenkogs replied:
Assume `fDisconnect` could be useful here too. Just in case `ForEachNode` gets updated.

mzumsande replied:
I also don't see that as a real risk, but I do see the problem is that whoever reads this code (or decides to use `ForEachNode` somewhere else) could easily get confused, especially since some callers check again for `fDisconnect`.
I added a refactor commit renaming `ForEachNode()` to `ForEachFullyConnectedNode()` and removing the duplicate checks / comments. What do you think about that approach?

naumenkogs replied:
To be clear, i suggested literally having `Assume(fDisconnect)`, not an extra `if` check (that would actually make code worse).

Even if you refactor it into `ForEachFullyConnectedNode`, I think this `Assume(fDisconnect)` is kinda useful — because who knows what exactly `ForEachFullyConnectedNode` means :)

But I'm fine either way, either you rename or add `Assume` (both are ideal i think :))

ajtowns commented on src/net.cpp:
```diff
@@ -2410,7 +2411,7 @@ int CConnman::GetExtraFullOutboundCount() const
         }
     }
-    return std::max(full_outbound_peers - m_max_outbound_full_relay, 0);
+    return full_outbound_peers - m_max_outbound_full_relay;
```
Negative numbers mean neither of those code blocks are executed; if it was `max{...,0}` then the "at zero" code block would be executed when it shouldn't be.

sr-gi replied:
Right, because we want to execute code for the case in which we are exactly at the limit, and being under the limit will be mapped to the same value. My bad.

I have to say though that, after this change, the function name becomes a bit confusing.

ajtowns replied:
> I have to say though that, after this change, the function name becomes a bit confusing.

I had a quick look at changing it to `return full_outbound_peers` and doing the comparison with `m_max_outbound_full_relay` in the caller. Would require exposing the latter value to net_processing, but otherwise seems like it works. Wasn't convinced it was enough of an improvement to bother though.

mzumsande replied:
I could add this if reviewers prefer it. I also thought about renaming, but couldn't really think of a better name than "GetExtraFullOutboundCount" that would imply negative return values.

mzumsande replied:
I like that name! I missed your comment in the most recent push, but will change it tomorrow!

mzumsande replied:
done now!

mzumsande commented on test/functional/test_framework/p2p.py:
```diff
@@ -368,13 +371,16 @@ def peer_connect(self, *args, services=P2P_SERVICES, send_version=True, **kwargs
     def peer_accept_connection(self, *args, services=P2P_SERVICES, **kwargs):
+        txrelay = True
+        if 'txrelay' in kwargs:
+            txrelay = kwargs.pop('txrelay')
```
done, thanks

mzumsande commented on test/functional/test_framework/p2p.py:
```diff
@@ -351,12 +351,15 @@ def __init__(self, support_addrv2=False, wtxidrelay=True):
+        if txrelay:
+            vt.relay = P2P_VERSION_RELAY
+        else:
+            vt.relay = 0
```
used the ternary, I think having a default arg is not ideal here because I'm not sure what should be the default and why.

mzumsande commented on test/functional/p2p_blocksonly.py:
```diff
@@ -78,6 +78,33 @@ def blocksonly_mode_tests(self):
+
+        self.log.info("Check that we evict full-relay outbound connections to blocksonly peers at the full outbound limit")
+        # Simulate the situation of being at the max limit of full-outbound connections by setting -maxconnections=1
+        self.restart_node(0, ["-maxconnections=1"])
```
done, added another connection that is not being disconnected.

mzumsande commented:
Pushed to address feedback.

I also removed a sub-test with a block-relay-only peer, there isn't really a reason to think that that peer would be disconnected anyway, and it didn't make much sense in that form because we weren't at any connection limit.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -5246,6 +5247,27 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
     }
+    // If we are at the max full outbound limit but not above, check if any peers
+    // don't support tx relay (they might be in -blocksonly mode) - if so,
+    // evict the newest of them so we can fill the slot with a tx-relaying outbound peer
```
nit: is `the newest` enforced by `ForEachNode` iterating? Perhaps make this requirement more clear? Perhaps even add an assert for `m_connected` before `node_to_evict = node->GetId();`

mzumsande replied:
It is enforced by the line
`if (!node_to_evict.has_value() || *node_to_evict < node->GetId()) {` (second condition), 
not through `m_connected` but by picking the eligible node with the highest NodeId. I'll add to the comment that "newest" means "highest `Node Id`", do you think that would be sufficient?

naumenkogs replied:
sounds good

naumenkogs commented on test/functional/p2p_blocksonly.py:
```diff
@@ -78,6 +78,31 @@ def blocksonly_mode_tests(self):
+        blocksonly_peer1.peer_disconnect()
+
+        self.log.info("Check that we don't evict full-relay outbound connections to blocksonly peers if we are blocksonly ourselves")
+        self.restart_node(0, ["-blocksonly", "-maxconnections=1"])
```
Not sure why anyone would do `maxconnections=1` in practice, but then if their firewall allows to find only one blocksonly node (and they are not in blocksonly), the behavior might be confusing: they will be frequently disconencted from this peer.

Perhaps, this could be logged if maxconnections=1

naumenkogs replied:
the idea was to log `if maxconnect is set to 1`, so at the startup (setting this argument) time.
I don't insist though.

mzumsande replied:
I'd agree with adding this warning, but yeah, this PR is not the best place for it.

naumenkogs commented:
ACK 840a022

Not sure if you noticed [this idea](https://github.com/bitcoin/bitcoin/pull/28538#discussion_r1417175382), might be a worthy precaution.

amitiuttarwar reviewed (commented):
code looks good to me. left a comment about the test that seems worth fixing, then I'm ready to ACK the patch.

trying to think about any potential negative network effects that could occur, but so far it seems fine to disconnect OB full relay peers that signal they don't want to receive transactions.

mzumsande committed:
refactor: rename ForEachNode and remove duplicate checks

Rename ForEachNode to ForEachFullyConnectedNode because it
has an inbuilt check that only executes the passed function
for fully connected nodes (i.e. if fSuccessfullyConnected is set
and fDisconnect isn't).

This means callers don't need to need to worry about that, so remove
those duplicate checks and related comments.

mzumsande commented on test/functional/p2p_blocksonly.py:
```diff
@@ -78,6 +78,31 @@ def blocksonly_mode_tests(self):
         self.nodes[0].disconnect_p2ps()
         self.generate(self.nodes[0], 1)
 
+        self.log.info("Check that we sustain full-relay outbound connections to blocksonly peers when not at the full oubound limit")
```
Good catch, I missed that we are already in blocksonly mode! I moved the added tests to `blocksonly_peer_tests` subtest as suggested.

amitiuttarwar reviewed (commented):
ACK d636e38d79a4c3950da91090b1f787163f11e24d

mzumsande commented on test/functional/p2p_blocksonly.py:
```diff
@@ -78,6 +79,34 @@ def blocksonly_mode_tests(self):
         self.nodes[0].disconnect_p2ps()
         self.generate(self.nodes[0], 1)
 
+    # Tests in which our peer is blocksonly
```
done

mzumsande commented:
[d636e38 ](https://github.com/bitcoin/bitcoin/commit/d636e38d79a4c3950da91090b1f787163f11e24d)to 2a17ac3:
Renamed `GetExtraFullOutboundCount` to `GetFullOutboundDelta` since it can now return negative values, plust small doc change in the test.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -5241,6 +5242,27 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
+    // evict the newest of them (highest Node Id) so we can fill the slot with a tx-relaying outbound peer
+    // This is not done if we are in -blocksonly mode ourselves or still in IBD,
+    // because then we don't care if our peer supports tx relay.
+    else if (full_outbound_delta == 0 && !m_opts.ignore_incoming_txs && !m_chainman.IsInitialBlockDownload()) {
```
1393bf67b26c000483c750258ca9d1daeab8a2cb

not super clear why we won't do this `if full_outbound_delta > 0`?
Maybe a comment explaining why this could be possible in the first place could help (probably above the first if block).
Even now it's hard to tell if it's possible to be over the limit most-of-the-time, let alone if we add some new complexity for being over the limit more frequently (for some security reason or whatever).

mzumsande replied:
Hmm, I think the main reason was to avoid interaction with the existing extra-full-outbound eviction.
Which raises the question whether there should be any interaction?
Should nodes that are protected from that (either because `m_protect` is true or because they are the only ones for their network) also be made exempt from this new disconnection due to not supporting tx relay?
I'm not really sure about that: It would be unfortunate if the non-tx-relaying outbound peer would be the only one with the best chain, but how realistic would such a scenario be?

sr-gi replied:
> not super clear why we won't do this if full_outbound_delta > 0?

> Which raises the question whether there should be any interaction?

I think this would create a weird interaction. The current approach of this PR is trying to make the 8 full-outbound connections relay transactions, given we already have blocks-only connections, so we should not waste full-outbound slots for this purpose. However, our full-outbound connection protection criteria is based on whether they have provided valid blocks with enough accumulated work and it has nothing to do with how good they are at providing transactions (there is also the recently added network diversity criteria, more about that later). 

This raises the question of what we value the most, having 8 full-outbound peers relaying transactions or our protected peers (and under what criteria). If the former is valued the most, we could merge this under `full_outbound_delta > 0` but pick a node not taking the protection criteria into account (which raises the question of how useful the protection is to start with). If the latter is prioritized, then we exclude the protected peers first and then run the eviction over the remaining. This slightly changes the goal of the PR, given we may not achieve 8 transaction relaying transactions. Finally, there are the additional protected peers for network diversity. Depending on how this is treated we may end up with an even smaller number of full-outbounds relaying transactions (`8 - 4 - n` where `n` represents the number of networks we support).

mzumsande replied:
Yes, I agree - I think that there are colliding goals, my thinking is as follows:
1) The `m_protect` protection attempts to prevent pathological situations where we wouldn't have peers with the honest chain (that shouldn't happen normally even without the protection, but would be very serious if they did)
2) The network-specific protection aims to be connected to all networks we support, to help the interconnection, and also to reduce the risk of eclipse attacks. Ideally, that means relaying both blocks and transactions from one network to another.
3) The added code attempts to make sure we have as many full-outbound peers actually participating in tx relay as possible, mostly for selfish reasons (e.g. privacy, fee estimation) 

These goals conflict. I'm pretty confident that 3) should take precedence over 2), because the network-specific protection isn't really about this particular peer, we just want any connection to that network, and would be happy to exchange a current peer that doesn't support tx-relay with a tx-relaying peer from that same network.
I'm kind of torn between 1) and 2). It would be easy to exempt non-tx-relaying `m_protect` peers from disconnection, but that would mean that we could end up with up to 4 non-tx-relaying full outbound peers permanently.

stratospher replied:
> It would be easy to exempt non-tx-relaying m_protect peers from disconnection, but that would mean that we could end up with up to 4 of these peers permanently.

[commit message](https://github.com/bitcoin/bitcoin/pull/11490/commits/5a6d00c6defc587e22c93e63029fdd538ce8858d) where `m_protect` was introduced says "we pick 4 of our outbound peers and do not subject them to this logic, to be more conservative. We don't wish to permit temporary network issues (or an attacker) to excessively disrupt network topology."

wondering how someone can disrupt network topology with this and if it's still relevant.

naumenkogs commented on test/functional/p2p_blocksonly.py:
```diff
@@ -78,6 +79,34 @@ def blocksonly_mode_tests(self):
+        blocksonly_peer.sync_with_ping()
+        blocksonly_peer.peer_disconnect()
+
+        self.log.info("Check that we evict full-relay outbound connections to blocksonly peers at the full outbound limit")
```
68769ff7c3252c7cfb8758459218979eacba7acb

You could expand this test by adding some irrelevant peers (e.g. ADDRFETCH), but no big deal.

mzumsande replied:
I tried it, but it turned out to be annoying because in the test we simulate being at the full outbound limit by artificially running with a lower `-maxconnections`. If we now also add an unrelated peer like ADDRFETCH , it would make the test fail because we don't have enough slots  in `semOutbound`.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2020,7 +2020,7 @@ void PeerManagerImpl::NewPoWValidBlock(const CBlockIndex *pindex, const std::sha
+    m_connman.ForEachFullyConnectedNode([this, pindex, &lazy_ser, &hashBlock](CNode* pnode) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
         AssertLockHeld(::cs_main);
 
         if (pnode->GetCommonVersion() < INVALID_CB_NO_BAN_VERSION || pnode->fDisconnect)
```
ad48667ec8e8d563550df768d0b45abf800662d9

gotta drop the second condition here?

mzumsande replied:
done, thanks - I had missed this!

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -5194,12 +5194,11 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
+            // Only consider outbound-full-relay peers
+            if (!pnode->IsFullOutboundConn()) return;
             CNodeState *state = State(pnode->GetId());
             if (state == nullptr) return; // shouldn't be possible, but just in case
```
ad48667ec8e8d563550df768d0b45abf800662d9

might update this too `if (!Assume(state)) return` and drop the comment if you feel like it :)

mzumsande replied:
I think I'd prefer to not do it here, for my taste it's not related enough to the core of this PR.

mzumsande commented:
Updated to address feedback, I would love to hear more opinions on the protection issue discussed in the threat following https://github.com/bitcoin/bitcoin/pull/28538#discussion_r1429773492

stratospher commented on test/functional/p2p_blocksonly.py:
```diff
@@ -78,6 +79,34 @@ def blocksonly_mode_tests(self):
+        # Have one blocksonly peer that is not being disconnected
+        blocksonly_peer1 = self.nodes[0].add_outbound_p2p_connection(P2PInterface(), p2p_idx=1, txrelay=False)
+        blocksonly_peer2 = self.nodes[0].add_outbound_p2p_connection(P2PInterface(), p2p_idx=2, txrelay=False)
+        self.nodes[0].mockscheduler(46)
```
68769ff: nit if you retouch.
```suggestion
with self.nodes[0].assert_debug_log(expected_msgs=["disconnecting full outbound peer not participating in tx relay: peer=1"]):
            self.nodes[0].mockscheduler(46)
            blocksonly_peer2.wait_for_disconnect()
```

stratospher reviewed (commented):
ACK 8f07458. useful to fill outbound slots with peers which we can relay transactions with and i liked the approach in this PR to swap away blocks-only peers in full relay outbound slots only after the threshold number of max full relay outbound connections is reached.

mzumsande commented:
Thanks for the review!

We had an offline conversation about this today (FYI @sdaftuar @sipa @sr-gi). Some takeaways:
- It would probaly be better to first have a replacement peer ready before disconnecting a blocksonly peer. That would mean a similar approach as for the extra network-specific peers, it would also mean that we would need to add a way to make extra connections when we have a -blocksonly peer we want to get rid of.
- We should avoid situations where there could be perpetual churn. For example, imagine the handful of peers on the cjdns network suddenly all went blocksonly. Then our network-specific logic would try to make new connections to them, while this logic would disconnect them, resulting in a lot of churn.
- To avoid interactions with protected peers, it might be best to not protect `fRelay=false` peers in the first place via `m_protect`.

I think I'll look into reworking the PR and will put it in draft until that has happened.

mzumsande commented:
Not working on this currently, I might open a new PR at a later time, or up for grabs if someone else wants to.
